### ISSUE TITLE
Significant performance degradation between 1.5.1 and 1.6.0+

### ISSUE THREAD

**Issue by zikaeroh** [jackc/pgconn#45]
Somewhere between 1.5.1 and 1.6.0, I'm seeing a pretty significant performance loss, mostly in memory allocations. Before, I was on `v1.5.1-0.20200408003821-5d2be99c254e`, which on one test would give me:

```
goos: linux
goarch: amd64
pkg: github.com/hortbot/hortbot/internal/bot
BenchmarkHandleNopParallel-4   	    4662	    252455 ns/op	    7962 B/op	      92 allocs/op
PASS
ok  	github.com/hortbot/hortbot/internal/bot	4.846s
```

But after upgrading, I get:

```
goos: linux
goarch: amd64
pkg: github.com/hortbot/hortbot/internal/bot
BenchmarkHandleNopParallel-4   	    4244	    266762 ns/op	   15069 B/op	     259 allocs/op
PASS
ok  	github.com/hortbot/hortbot/internal/bot	4.711s
```

Bisecting is a bit difficult because PRs here aren't squashed merged, but attempting to do it by hand shows that #36 was the point at which the performance changed.

Before, my pprof (allocation count) would look like:

![image](https://user-images.githubusercontent.com/48577114/85944517-2e7da400-b8ec-11ea-8e2d-357f7f19fa77.png)

But after, a lot of the memory allocations are performed elsewhere:

![image](https://user-images.githubusercontent.com/48577114/85944508-14dc5c80-b8ec-11ea-8d28-a81910d4f8ae.png)

I've tested this in my project (https://github.com/hortbot/hortbot) by comparing the latest `pgconn` with what I get after doing `go get -d github.com/jackc/pgconn@5d2be99c254e`.

(I realize this could be more to do with pgx than pgconn, but I filed it here given downgrading pgconn only showed a difference.)

**Comment by zikaeroh** [comment]
I've traced this down, and it is an interplay between #36, pgx, and `database/sql`. `Columns` is a hot function, but the implementation in `pgx` makes a copy of the columns names each time. Before #36, the descriptions would have been set to `nil`, so sometimes it would return no names after the command was concluded. This appeared to not have any ill effect, at least not in pgx or its stdlib driver. When the fix was applied, now the calls that did nothing now also copied things.

The fix will be to `pgx` as an optimization to the `Columns` function to cache its result (and reset when the field description slice changes).

With the fix I will submit for this and jackc/pgx#783 together, the regression goes away and the `nextLock` path is an order of magnitude better.

![image](https://user-images.githubusercontent.com/48577114/85960342-c369b700-b957-11ea-8ff0-5eebe389559c.png)

I'll close this after I submit my PR to pgx.

**Comment by zikaeroh** [comment]
Sent jackc/pgx#784.
